### PR TITLE
Add multientities single batch request to aerospike template

### PR DESCRIPTION
**Pull request by mrozk**

### CODE REVIEW

File: src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java [hunk 1]
```diff
@@ -353,6 +355,37 @@ private Record getAndTouch(Key key, int expiration) {
 		}
 	}
 
+	@Override
+	public Map<Class<?>, Collection<?>> findEntitiesByIds(Map<Class<?>, Collection<?>> entitiesKeys) {
+		Assert.notNull(entitiesKeys, "Map of entities ids must not be null!");
```
Comment (reugn) [inline src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java:360]: @mrozk why not
```java
return entitiesKeys.entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey,
				e -> findByIdsInternal(e.getValue(), e.getKey())));
```
which makes this method redundant.

**Review by reugn — CHANGES_REQUESTED** [review]

File: src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java (hunk 1 above)
Comment (mrozk) [reply src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java:360]: In your scenario you will make N batch requests, this method does 1 batch request

File: src/main/java/org/springframework/data/aerospike/core/AerospikeOperations.java [hunk 2]
```diff
@@ -110,6 +111,8 @@
 
 	<T> List<T> findByIds(Iterable<?> ids, Class<T> entityClass);
 
+	Map<Class<?>, Collection<?>> findEntitiesByIds(Map<Class<?>, Collection<?>> keysMap);
```
Comment (Aloren) [inline src/main/java/org/springframework/data/aerospike/core/AerospikeOperations.java:114]: please rename to `findByIds`.
Also it would be nice to have javadoc about tis method

File: src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java [hunk 3]
```diff
@@ -691,4 +724,16 @@ private Node getRandomNode(Node[] nodes) {
 		}
 		throw new AerospikeException.InvalidNode("Command failed because no active nodes found.");
 	}
+
+	private <T> List<Key> toKeysList(Class<T> entityClass, Iterable<?> ids) {
```
Comment (Aloren) [inline src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java:728]: `ids` has list type in the public method. what is the reason for it to be iterable here?

File: src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java [hunk 4]
```diff
@@ -691,4 +724,16 @@ private Node getRandomNode(Node[] nodes) {
...
+		Assert.notNull(entityClass, "List of ids must not be null!");
```
Comment (Aloren) [inline src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java:729]: message is invalid. please fix

File: src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java [hunk 5]
```diff
@@ -353,6 +355,37 @@ private Record getAndTouch(Key key, int expiration) {
...
+	private Map<Class<?>, List<Object>> findEntitiesByIdsInternal(Map<Class<?>, List<Key>> entitiesKeys) {
+		Class<?>[] entityClasses = entitiesKeys.entrySet().stream()
+				.flatMap(entry -> entry.getValue().stream().map(item -> entry.getKey()))
+				.toArray(Class<?>[]::new);
+
+		Key[] keys = entitiesKeys.entrySet().stream()
+				.flatMap(entry -> entry.getValue().stream())
+				.toArray(Key[]::new);
+
+		Record[] records = client.get(null, keys);
```
Comment (Aloren) [inline src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java:380]: if request is empty, i think, we can skip requesting anything from the client?

File: src/test/java/org/springframework/data/aerospike/core/AerospikeTemplateFindByEntitiesTests.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        expectedResult.put(Person.class, persons);
+        expectedResult.put(Customer.class, emptyList());
+
+        assertThat(entitiesByIds).usingRecursiveComparison()
+                .ignoringCollectionOrder()
+                .isEqualTo(expectedResult);
+    }
+
+    @Test
+    public void shouldReturnMapWithEmptyResultsOnEmptyRequest() {
```
Comment (Aloren) [inline src/test/java/org/springframework/data/aerospike/core/AerospikeTemplateFindByEntitiesTests.java:54]: what's the difference from `shouldReturnMapWithEmptyResults` ?

File: src/test/java/org/springframework/data/aerospike/core/AerospikeTemplateFindByEntitiesTests.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+        Map<Class<?>, Collection<?>> expectedResult = new HashMap<>();
+        expectedResult.put(Person.class, persons);
+        expectedResult.put(Customer.class, customers);
+
+        assertThat(entitiesByIds).usingRecursiveComparison()
+                .ignoringCollectionOrder()
+                .isEqualTo(expectedResult);
+    }
+
```
Comment (Aloren) [inline src/test/java/org/springframework/data/aerospike/core/AerospikeTemplateFindByEntitiesTests.java:100]: pls add tests for requesting bunch of keys where some of them are missing, all of them are missing. probably some other corner cases as well

File: src/main/java/org/springframework/data/aerospike/core/AerospikeOperations.java (hunk 2 above)
Comment (reugn) [reply src/main/java/org/springframework/data/aerospike/core/AerospikeOperations.java:114]: @Aloren I find this method's signature vague.

File: src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java (hunk 1 above)
Comment (reugn) [reply src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java:360]: Have you benchmarked both?

Comment (mrozk) [reply src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java:360]: I suppose that batches for different set work the same as batches to single set. Also if we cover every request in spring retry we will receive profit from requests count perspective. We have use cases where  we don't want to send 2 request because it is too expensive - timeout 50ms + 3 retries on each (if aerospike is glitching)

File: src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java (hunk 3 above)
Comment (mrozk) [reply src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java]: Fixed to Collection

File: src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java (hunk 4 above)
Comment (mrozk) [reply src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java]: done

File: src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java (hunk 5 above)
Comment (mrozk) [reply src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java:385]: Added check on the beginning of the method

File: src/test/java/org/springframework/data/aerospike/core/AerospikeTemplateFindByEntitiesTests.java (hunk 6 above)
Comment (mrozk) [reply src/test/java/org/springframework/data/aerospike/core/AerospikeTemplateFindByEntitiesTests.java:110]: I have renamed methods to clarify their functionality

File: src/test/java/org/springframework/data/aerospike/core/AerospikeTemplateFindByEntitiesTests.java (hunk 7 above)
Comment (mrozk) [reply src/test/java/org/springframework/data/aerospike/core/AerospikeTemplateFindByEntitiesTests.java:174]: done

**Review by reugn — COMMENTED** [review]
This method has a number of code smells:
* Use of wildcards in return types.
* The type of the parameter is the same as the return type and wildcards make it ambiguous.
* The method's name doesn't add clarification.

The API user will not easily understand what this method does.
I would suggest returning something like:
```java
@FunctionalInterface
interface GroupedEntities {
    <T> List<T> getByClass(Class<T> entityClass);
}
```

**Comment by Aloren** [comment]
@mrozk did you have a chance to update your PR with the changes requested by @reugn ? this will definitely improve the method readability significantly

**Comment by mrozk** [comment]
> @mrozk did you have a chance to update your PR with the changes requested by @reugn ? this will definitely improve the method readability significantly

@Aloren I'm going to update the PR today or tomorrow with suggested changes

File: src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java [hunk 8]
```diff
@@ -334,6 +355,53 @@ private Record getAndTouch(Key key, int expiration) {
 		}
 	}
 
+
+	/**
+	 * Executes a single batch request to get results for several entities.
+	 *
+	 * Aerospike provides functionality to get records from different sets in 1 batch
+	 * request. The methods allows to put grouped keys by entity type as parameter and
+	 * to get result as spring data arospike entities grouped by entity type.
```
Comment (reugn) [inline src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java:364]: typo

File: src/main/java/org/springframework/data/aerospike/core/model/GroupedKeys.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
+package org.springframework.data.aerospike.core.model;
```
Comment (reugn) [inline src/main/java/org/springframework/data/aerospike/core/model/GroupedKeys.java:1]: Please add a copyright header.

File: src/main/java/org/springframework/data/aerospike/core/model/GroupedEntities.java [hunk 10]
```diff
@@ -0,0 +1,51 @@
+package org.springframework.data.aerospike.core.model;
```
Comment (reugn) [inline src/main/java/org/springframework/data/aerospike/core/model/GroupedEntities.java:1]: Please add a copyright header.

**Review by reugn — CHANGES_REQUESTED** [review]
@mrozk LGTM.
Please add this functionality to `ReactiveAerospikeOperations`.

**Comment by reugn** [comment]
@mrozk what are your plans regarding this PR?

File: src/main/java/org/springframework/data/aerospike/core/model/GroupedEntities.java (hunk 10 above)
Comment (mrozk) [reply src/main/java/org/springframework/data/aerospike/core/model/GroupedEntities.java:1]: done

File: src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java (hunk 8 above)
Comment (mrozk) [reply src/main/java/org/springframework/data/aerospike/core/AerospikeTemplate.java:364]: done

**Comment by mrozk** [comment]
> @mrozk what are your plans regarding this PR?

@reugn I'm going to finish reactive implementation this week. We can either wait for it and release both implementation or release sync version first and release async separately.

**Comment by roimenashe** [comment]
@mrozk any updates regarding the async functionality?

**Comment by mrozk** [comment]
> @mrozk any updates regarding the async functionality?

Hi! done

**Comment by reugn** [comment]
Thank you @mrozk .

**Review by reugn — APPROVED** [review]
